src/hardware/sblaster.cpp

```diff
@@ -1973,38 +1974,72 @@ static uint8_t CTMIXER_Read() {
 	return ret;
 }
 
+// Called by DSP_WRITE_STATUS to check if the write buffer is at capacity.
+static bool write_buffer_at_capacity()
+{
+	// Is the DSP in an abnormal state and therefore we should consider the
+	// buffer at capacity and unable to receive data?
+	if (sb.dsp.state != DSP_S_NORMAL) {
+		return true;
+	}
+	// Report the buffer as having some room every 8th call, as the buffer
+	// will have run down by some amount after sequential calls.
+	if ((++sb.dsp.write_status_counter % 8) == 0) {
+		return false;
+	}
+	// If DMA isn't running then the buffer's definitely not at capacity.
+	if (sb.dma.mode == DSP_DMA_NONE) {
+		return false;
+	}
+	// Finally, the DMA buffer is considered full until it's able to accept
+	// a full (64 KB) write; which is once we've hit the minimum threshold.
+	return (sb.dma.left > sb.dma.min);
+}
+
+// Sound Blaster DSP status byte
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Ref: http://archive.gamedev.net/archive/reference/articles/article443.html
+//
+// Read from 02x0Ch (DSP Write Buffer Status): Bit 7, when 0, indicates that the
+// DSP is ready to receive data through the DSP Write Data or Command port
+// (02x0Ch write).
+//
+// Read from 02x0Eh (DSP Data Available Status): Bit 7, when 1, indicates that
+// the DSP has pending data to be read through the DSP Read Data port (02x0Ah
+// read).
+//
+union BufferStatus {
+	uint8_t data = 0;
+	bit_view<7, 1> has_data;
+	// when 1, the buffer has data and is ready to send
+	// when 0, the buffer doesn't have data and is ready to receive
+};
+
 static uint8_t read_sb(io_port_t port, io_width_t)
 {
 	switch (port - sb.hw.base) {
 	case MIXER_INDEX: return sb.mixer.index;
 	case MIXER_DATA: return CTMIXER_Read();
 	case DSP_READ_DATA: return DSP_ReadData();
-	case DSP_READ_STATUS:
-		//TODO See for high speed dma :)
-		if (sb.irq.pending_8bit)  {
-			sb.irq.pending_8bit=false;
+	case DSP_READ_STATUS: {
+		// TODO See for high speed dma :)
+		if (sb.irq.pending_8bit) {
+			sb.irq.pending_8bit = false;
 			PIC_DeActivateIRQ(sb.hw.irq);
 		}
-		if (sb.dsp.out.used) return 0xff;
-		else return 0x7f;
-	case DSP_ACK_16BIT:
-		sb.irq.pending_16bit=false;
-		break;
-	case DSP_WRITE_STATUS:
-		switch (sb.dsp.state) {
-		case DSP_S_NORMAL:
-			sb.dsp.write_busy++;
-			if (sb.dsp.write_busy & 8) return 0xff;
-			return 0x7f;
-		case DSP_S_RESET:
-		case DSP_S_RESET_WAIT:
-			return 0xff;
-		}
-		return 0xff;
-	case DSP_RESET:
-		return 0xff;
+		BufferStatus read_status = {};
+		read_status.has_data = (sb.dsp.out.used != 0);
+		return read_status.data;
+	}
+	case DSP_ACK_16BIT: sb.irq.pending_16bit = false; break;
+	case DSP_WRITE_STATUS: {
+		BufferStatus write_status = {};
+		write_status.has_data = write_buffer_at_capacity();
+		return write_status.data;
+	}
+	case DSP_RESET: return 0xff;
```

Comment:
I'd do this here, more self-documenting:

```cpp
BufferStatus status = {};
status.flip();
return status;
```

(Also, without setting the low 7 bits to 1 as suggested, this is a discrepancy because the low 7 bits _are_ set in the reset status response.)

Reply:
(default high bits will take care of it)